restapi/utils_test.go

```diff
@@ -0,0 +1,219 @@
+func TestPerformReadRequestProperResponse(t *testing.T) {
+	// start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
+		// send response to be tested later
+		responseWriter.Write([]byte(`OK`))
+	}))
+	// close the server when test finishes
+	defer server.Close()
+
+	// try to read response from the server
+	body, err := restapi.PerformReadRequest(server.URL)
+	if err != nil {
+		t.Fatal("Unable to perform request", err)
+	}
+
+	// check for proper body
+	if !bytes.Equal(body, []byte(`OK`)) {
+		t.Fatal("Improper response body", err)
+	}
+}
+
+// TestPerformReadRequestStatusCode check how response can be processed by performReadRequest function.
+func TestPerformReadRequestStatusCode(t *testing.T) {
+	// start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
+		// send response to be tested later
+		responseWriter.WriteHeader(http.StatusInternalServerError)
+		responseWriter.Write([]byte(`error`))
+	}))
+	// close the server when test finishes
+	defer server.Close()
+
+	// try to read response from the server
+	body, err := restapi.PerformReadRequest(server.URL)
+	if err == nil {
+		t.Fatal("Error is expected")
+	}
+
+	// body needs to be nil in case of any error
+	if body != nil {
+		t.Fatal("Body needs to be nil in case of any error")
+	}
+}
+
+// TestPerformReadRequestEmptyBody check how response can be processed by performReadRequest function.
+func TestPerformReadRequestEmptyBody(t *testing.T) {
+	// start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
+		// send response to be tested later
+		responseWriter.WriteHeader(http.StatusOK)
+	}))
+	// close the server when test finishes
+	defer server.Close()
+
+	// try to read response from the server
+	body, err := restapi.PerformReadRequest(server.URL)
+	if err != nil {
+		t.Fatal("Unable to perform request", err)
+	}
+
+	// check for proper body - it needs to be empty this time
+	if len(body) > 0 {
+		t.Fatal("Improper response body", err)
+	}
+}
+
+// TestPerformReadRequestErrorInCommunication check how response can be processed by performReadRequest function.
+func TestPerformReadRequestErrorInCommunication(t *testing.T) {
+	// try to read response from the server, but by using improper URL
+	body, err := restapi.PerformReadRequest("")
+	if err == nil {
+		t.Fatal("Error is expected")
+	}
+
+	// body needs to be nil in case of any error
+	if body != nil {
+		t.Fatal("Body needs to be nil in case of any error")
+	}
+}
+
+// TestPerformWriteRequestProperResponse check the behaviour of function performWriteRequest
+func TestPerformWriteRequestProperResponse(t *testing.T) {
+	// start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
```

Comment:
Maybe we could extract that peace of code to function, because it's repeat across the file?

Reply:
do you mean the HTTP server constructor? Because the handler is different from test to test.

Reply:
Yes, but we can for example make one function with `httptest.NewServer(http.HandlerFunc`, and put as argument function handler

Reply:
ok good observation

Reply:
I faced the same issue and I thought of only one solution... Let me know what you think.
Since the repeating part is creating the test server and the insides of the anonymous Handler function actually do change, the only solution I thought of was to have one server and multiple handler functions and therefore multiple endpoints, right? I thought that this solution added more complexity than it removed.

Please correct me if I'm wrong.

Reply:
@Bee-lee I think that @bond wanted something similar to https://github.com/RedHatInsights/insights-operator-cli/pull/91/commits/bb987dbcde027d75affbc839660801d31c7bee5e

WDYT now @bond95

Reply:
Ahh, ok, I see now.